### ISSUE TITLE
Work with official QT

### ISSUE THREAD

**Issue by ItachiSan** [telegramdesktop/tdesktop#1815]
This issue is opened in order to avoid in future the usage of a patched QT version.
The general idea would be to:
- Use system FCTIX QT plugin as many Linux distros can offer this plugin, else it can be compiled as dependency 
- Avoid patching of QT stuff or send stuff upstream, in order to have a pure environment
  I would like to have feedback from developers and in case upload some code to patch stuff.

**Comment by john-preston** [comment]
1. as I understand I can't use it while linking Qt statically: "Note that the QPluginLoader cannot be used if your application is statically linked against Qt. In this case, you will also have to link to plugins statically"

2. no way to link dynamically to Qt (even if you avoid all the patches I use), because my text rendering code is based on the internal code of Qt, not the public interface provided, so I can't rely that the internal methods still exist and do the expected thing.

**Comment by YtvwlD** [comment]
I think the idea would be to link dynamically.
I don't really understand what's the problem here. I already have QT installed on my system. Why can't Telegram use it?

**Comment by ItachiSan** [comment]
@john-preston can't be the Telegram Desktop code refactored to use the official text rendering methods...?

@YtvwlD because Telegram Desktop use a patched version of QT5, so the methods in the official QT5 are different to the ones present in the patched version.

**Comment by jhasse** [comment]
This would also fix that the file dialog doesn't use my theme:

![bildschirmfoto vom 2016-06-17 10-26-49](https://cloud.githubusercontent.com/assets/80071/16148906/02663752-348d-11e6-9270-484b8cde704e.png)

**Comment by YtvwlD** [comment]
Currently, nothing uses the system theme.
![spectacle d13590](https://cloud.githubusercontent.com/assets/5578100/16150868/1a9edf42-349a-11e6-80bc-9b5f19eb1602.png)

**Comment by AndydeCleyre** [comment]
@john-preston 

> my text rendering code is based on the internal code of Qt, not the public interface provided

What's the big advantage here? Lots of users and package maintainers would love for this to use official Qt, dynamically linked. So what is so advantageous about relying on patches and unreliable internal code that makes this worth it?

**Comment by john-preston** [comment]
@AndydeCleyre Well, when it was first made, I was working on a fast holding in memory + rendering of big amounts of text parts with graphic emoji, so that one message history can hold tens of thousands of text messages (I've tested up to 100 000) and window resize (which triggers the resize of all the messages to count the new window height) should work without lagging and scrolling should be smooth with instant switches between different conversations with big amounts of messages (tdesktop holds all the loaded messages in memory, local database for messages is not yet supported, so it needs all that) with acceptable memory consuming. I don't know how to achieve this using just Qt public interface.

**Comment by plinnell** [comment]
I can tell you as a developer for a project written in Qt since 2001, the patching of Qt by an  application is a **very** bad idea.  See: [http:www.scribus.net](url) We also deal with text rendering issues and have our own canvas because of the need to manipulate objects beyond text. 

As a long time Linux user (15 years) and having packaged Linux applications for more than 10, not using Distro provided Qt over time is going to be an unholy, difficult to maintain mess.  What happens when the patched Qt needs a security patch ? and Qt does get them ;) **You are creating your own technical debt** and a relatively young project needs to avoid this especially in the early days.  

 It seems most of what you are trying to fix, could be fixed by pulling in fcitx-qt5-devel, which provides the input capabilities for CJK.fcitx is pretty well supported on most recent distros.

Requiring a patched Qt is a **very** strong barrier to get your application into Linux distros. I'm one of the Factory maintainers for openSUSE and this would more than likely be rejected to be part of the distro, especially statically linking Qt and/or having your own patched Qt. I'm pretty certain Fedora/RedHat would say the same thing. Moreover, even if you apply patches, separating them out to different platforms is a best practice. 

Now enough ranting. I do hope your team considers using [https://openbuildservice.org](url) to do your Linux packaging. It is cross distro and has many automation features which can ease package builds **and** provides infra for distribution.  It can be hooked into Jenkins and speaks native git for pulling sources to build.

**Comment by gavv** [comment]
I totally agree with @plinnell. Also, patched Qt was the main reason why telegram source package was [rejected](https://github.com/gentoo/gentoo/pull/1551) by Gentoo developers. Bundled dependencies are usually very unwelcome by any distro maintainers and Linux community.

**Comment by hobarrera** [comment]
@plinnell @gavv I don't think there's any rejection to actually stop patching Qt.

As far as I've understood, it's just that the devs haven't been able to get things working as expected with the upstream Qt, and nobody's offered another working solution (in the form of a PR). So rather than "up for debate", I think this issue is actually "awaiting patches".

**Comment by guoyunhe** [comment]
@gavv Some people already submited some pull request to use existing fcitx-qt5-devel package rather than patch the whole fcitx binding. But these pull requests have never  been accepted.

**Comment by john-preston** [comment]
@guoyunhe fcitx is the small thing in the patch, it just got there because the patch was already in place.

The main problem is the text-with-emoji recounting/rendering, which uses private Qt apis.

I see no way in not using it trying to make fast-enough and not-so-memory-consuming text-with-emoji processing for tdesktop. And even if there is such way, which I don't know, it will be a huge amount of work to move the existing code base to that another way.

**Comment by AndydeCleyre** [comment]
@john-preston is it relatively simple to make a branch that uses Qt's slow/mem-heavy methods? How bad is it?

**Comment by john-preston** [comment]
@AndydeCleyre I don't know, how to make that in an easy way. You need to move all the SourceFiles/ui/text/text_.. on something completely different, that uses only public Qt API.

**Comment by john-preston** [comment]
@AndydeCleyre I guess you need to create a separate QTextDocument object for each displayed text piece and wrap all the methods that work with Text() right now to work with those documents.

**Comment by pschichtel** [comment]
What about upstreaming the changes? If your controls are in fact that much more efficient than the standard Qt ones, I guess a lot of other applications could benefit from them as well.

**Comment by john-preston** [comment]
@pschichtel I'm pretty sure this can't be done :( They're not general controls in any way, just very app-specific pieces of code. Perhaps the approach that was used there could be generalized some how but it is a huge amount of work in its own and after that I still doubt that any of it will be merged to Qt.

**Comment by pschichtel** [comment]
@john-preston Are entire controls implemented in the Qt patches? I through more about the optimizations you did. Or you could add in public APIs for the private ones you are using and upstream them.
This all sounds like hacking around limitations of a library instead of just removing the limitations from the library.

**Comment by hobarrera** [comment]
Is is somehow possible to make those "private Qt APIs" public upstream? Has this even been discussed?

**Comment by AndydeCleyre** [comment]
This is by far the most important bug in this project, and I would contribute to a bug bounty for it.

**Comment by Schekolda** [comment]
So, will you stop using patched QT?

**Comment by stek29** [comment]
https://aur.archlinux.org/packages/telegram-desktop-systemqt-notoemoji/ may help (from #3129)

**Comment by KireinaHoro** [comment]
As well as https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=telegram-desktop-systemqt .

**Comment by AndydeCleyre** [comment]
Can someone write here what the deal is with those "systemqt" packages? What works and what doesn't?

**Comment by emanuelserpa** [comment]
It's working great here, I'm not seeing anything not working in https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=telegram-desktop-systemqt

**Comment by AndydeCleyre** [comment]
Wow, it does seem to work perfectly so far. @john-preston can you check this out and point out what we should expect to be broken -- and if nothing (or very little) is, how this might affect the dependency on your custom Qt going forward?

**Comment by emanuelserpa** [comment]
This is a very important issue and the only reason telegram isn't packaged available in the linux distributions' official repositories.

**Comment by agur4ik** [comment]
By the way, using system qt fixes this ancient bug: #1026. It appears to be libappindicator bug, I guess, as systemqt version (https://aur.archlinux.org/packages/telegram-desktop-systemqt) uses native tray icon.

**Comment by jhasse** [comment]
I don't think that has anything to do with Qt, but if the build uses libappindicator or not.

**Comment by emanuelserpa** [comment]
Any updates?

**Comment by portaloffreedom** [comment]
Not using official Qt actually breaks the program running over wayland

**Comment by orschiro** [comment]
@agur4ik what you were saying is related to this theme inconsistency, correct?

![image](https://user-images.githubusercontent.com/170947/34243244-21e9227e-e620-11e7-821a-b03f11c7b107.png)

**Comment by mxa** [comment]
I suspect that fixing this would also resolve two other bugs:

1. make nimf work with Telegram https://github.com/telegramdesktop/tdesktop/issues/2603
2. compose key currently broken in Telegram https://github.com/telegramdesktop/tdesktop/issues/2603

**Comment by agur4ik** [comment]
@orschiro, it might be related but I'm not sure as I've not tested that. But since you are using GTK-based environment, theme inconsistency more probably has something to do with different GTK and Qt styles. Here's described how to fix it: https://wiki.archlinux.org/index.php/Uniform_look_for_Qt_and_GTK_applications. 
(I hope I got your question right)

**Comment by Peque** [comment]
@john-preston Friendly ping. :innocent:

**Comment by john-preston** [comment]
@Peque You can read my replies above. I don't know an easy and reliable way of moving this project to the non-custom Qt. Most of the patching is used on Windows and OS X, but some on Linux as well. Right now:

— Text layout uses internal Qt API.. All attempts to work with official Qt up till now still used it, just hoped that the required methods are being present in the dynamic libraries. If Qt some day changes or removes those methods the app would simply fail to launch.

— Open Sans Semibold font that is used widely in the app wouldn't load without patching Qt for that font. I don't know whose fault is that, but I couldn't make it work, Qt saw it just as another "Open Sans" font and all semibold text parts were rendered as normal text (names in chats list etc).

— Detection of the compositing manager was backported from next versions of Qt, so I could know if transparent windows are supported and not try to show them if they aren't.

Fixing bugs only in the upstream Qt is inefficient. If you see bugs / crash reports and you manage to find and fix them in Qt and even if you manage to report and include those fixes in some next version (which requires huge amount of work, I almost never have time for this), your current users still won't be able to use your app.

Also I'll need to deploy current versions anyway, because all current users are using the app and autoupdating to the next versions, if I just remove the linked Qt in some cases the app just won't launch anymore.

**Comment by jhasse** [comment]
What about using https://github.com/CrimsonAS/gtkplatform/ for your static Qt version? This way the theming would be picked up from system Gtk at least.

**Comment by orschiro** [comment]
@agur4ik

>  it might be related but I'm not sure as I've not tested that. But since you are using GTK-based environment, theme inconsistency more probably has something to do with different GTK and Qt styles. Here's described how to fix it: https://wiki.archlinux.org/index.php/Uniform_look_for_Qt_and_GTK_applications.
(I hope I got your question right)

You got it right! However, I tried this already. If I understand correctly, then this would work if Telegram Desktop uses the non-forked version of Qt. 

@jhasse 

> What about using https://github.com/CrimsonAS/gtkplatform/ for your static Qt version? This way the theming would be picked up from system Gtk at least.

Does the user have to use this or the developer?

**Comment by Peque** [comment]
@john-preston Thanks for your reply. Even though I read your comments above, it is good to have an update on this matter after more than one year. Just in case anything had changed. :wink:

**Comment by emanuelserpa** [comment]
@john-preston What's your thought about https://www.archlinux.org/packages/community/x86_64/telegram-desktop/? They are using systemqt now.

**Comment by john-preston** [comment]
@emanuelserpa This looks promising. If I understood correctly they've just copied small amount of code from Qt source that isn't available through Qt public API and compile it inside the binary.

**Comment by agur4ik** [comment]
@emanuelserpa I think, in Arch it is easy to incorporate system Qt Telegram build as it uses the latest and greatest software (@john-preston pointed out that some fuctionality is backported from the newer versions of Qt). Distributing Telegram with native Qt would result in the need to provide different TG builds for the different Qt versions. I never thought about this earlier though. Now I think that the issue should be resolved in downstream rather than upstream: let distro maintainers build TG with native Qt (if they will) and TG devs continue providing static build unified for all distros. However, it might be a good idea to simplify building the app against the different Qt versions by making this thing configurable (if that's easily possible) in order to avoid patching from distro maintainers. 

P.S.: sorry for my English.

**Comment by stek29** [comment]
> Distributing Telegram with native Qt would result in the need to provide different TG builds for the different Qt versions

It shouldn't. At least when done properly.

**Comment by orschiro** [comment]
>  let distro maintainers build TG with native Qt (if they will) and TG devs continue providing static build unified for all distros

Let's take Ubuntu as an example. It doesn't provide Telegram Desktop in its official repositories. But as far as I know there is a Snap as well as Flatpak for TD. Does this make a difference to the argument?

**Comment by gasinvein** [comment]
Use of patched Qt is a strong argument *against* adding Telegram to official repositories. Maybe if Telegram would work with distro's native Qt, there will be no need in Snap or Flatpak for it.

**Comment by loimu** [comment]
>Let's take Ubuntu as an example. It doesn't provide Telegram Desktop in its official repositories.

Really? [What's that then?](https://packages.ubuntu.com/search?keywords=telegram-desktop&searchon=names&suite=all&section=all)

**Comment by orschiro** [comment]
@loimu forget what I just said. Don't even know how I missed that. And I was sure it didn't have the package the last time I was looking for it in the official repos. :sweat_smile:

**Comment by shiba89** [comment]
ebuilds for Gentoo are available too https://github.com/reagentoo/gentoo-overlay/tree/master/net-im/telegram-desktop
I've been using this for a couple of weeks and not a problem so far.

**Comment by gasinvein** [comment]
So, as for now, [Fedora (rpmfusion)](https://pkgs.rpmfusion.org/cgit/free/telegram-desktop.git/tree/telegram-desktop-build-fixes.patch), [openSUSE](https://build.opensuse.org/package/view_file/server:messaging/telegram-desktop/tdesktop.patch), [Arch](https://git.archlinux.org/svntogit/community.git/tree/trunk/tdesktop.patch?h=packages/telegram-desktop), [Gentoo](https://github.com/reagentoo/gentoo-overlay/blob/master/net-im/telegram-desktop/files/patches/add-private-qt-functions.patch), [Debian/Ubuntu](https://salsa.debian.org/debian/telegram-desktop/blob/debian/master/debian/qt_functions.cpp) and even [Flathub](https://github.com/flathub/org.telegram.desktop/blob/master/Avoid-depending-on-static-libraries.patch), all patch Telegram to avoid linking statically with modified Qt. Did I miss someone?

**Comment by balsoft** [comment]
> So, as for now, [Fedora (rpmfusion)](https://pkgs.rpmfusion.org/cgit/free/telegram-desktop.git/tree/telegram-desktop-build-fixes.patch), [openSUSE](https://build.opensuse.org/package/view_file/server:messaging/telegram-desktop/tdesktop.patch), [Arch](https://git.archlinux.org/svntogit/community.git/tree/trunk/tdesktop.patch?h=packages/telegram-desktop), [Gentoo](https://github.com/reagentoo/gentoo-overlay/blob/master/net-im/telegram-desktop/files/patches/add-private-qt-functions.patch), [Debian/Ubuntu](https://salsa.debian.org/debian/telegram-desktop/blob/debian/master/debian/qt_functions.cpp) and even [Flathub](https://github.com/flathub/org.telegram.desktop/blob/master/Avoid-depending-on-static-libraries.patch), all patch Telegram to avoid linking statically with modified Qt. Did I miss someone?

You missed NixOS :smile:

**Comment by mikken** [comment]
Gentoo only has statically linked official binary package for Telegram Desktop.
The one above is a third-party package.

**Comment by ilya-fedin** [comment]
I'm not sure if this issue is valid, since dropping patches wouldn't have the effect that most of people want here: to use system Qt by the official binary.

1. Official binary supports Ubuntu 12.04+, what means that official binary will should link to Qt with version that is provided by that distro. I'm not sure if it even provides Qt 5, but if it provides, it would be something like 5.0.0, which will lack most of Qt APIs used by tdesktop
2. Telegram will lost one of it's main pros: just unpack binary and run it. Users will need to find Qt libraries in their repos (which are named very differenly across distributions and splitted differently too), check that they're new enough, etc, etc.

And most of Qt apps that provides official binaries doesn't use system Qt in their builds too: Viber, WPS Office, XnView MP, Calibre

**Comment by emanuelserpa** [comment]
I think that the issue was about compiling the Telegram's Desktop (not about the official binary) code to work in the distributions with the official/main qt and Telegram used a modified qt to compile and to compile it we needed patches like in https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=telegram-desktop-systemqt (that later become the official telegram-desktop package).

Now every distribution has telegram in the repositories and, at least here on Arch Linux, I don't see any patch in the PKGBUILD.

I don't know if I'm right, though.

**Comment by portaloffreedom** [comment]
@ilya-fedin You got this completely wrong: the issue is not with the binary distribution but with the possibility to compile the binary with the Qt dynamic libraries. Telegram was (still is? I don't know) doing weird and unnecessary modification to the Qt library and it made very difficult for distribution to include telegram in their own distro while removing unnecessary duplicated libraries.

Also, the number of people on linux that download precompiled open source programs from websites is close to zero.
Also, I would telegram-desktop to be 46MB, not 400MB.
Also, wayland support.

To sum up: the official binary can include the Qt libraries, I don't think people have problems with that. But please don't fork Qt for useless features so that it would only make the application 10 times bigger in our computers for the sake of it, while removing compatibility for new stuff e.g. with new wayland.

Make your life easier, leave Qt development to upstream.

**Comment by ilya-fedin** [comment]
> I think that the issue was about compiling the Telegram's Desktop

tdesktop imported distropatches in January, it can be compiled with system Qt of most of distros

**Comment by ilya-fedin** [comment]
> You got this completely wrong: the issue is not with the binary distribution but with the possibility to compile the binary with the Qt dynamic libraries.

Maybe some people here talking about compiling in distributions, but I see that the first post is about using system Qt in official bianry to work with fcitx, so that I assume that issue is about the official binary (anyway distribution part is not valid anymore)

**Comment by ilya-fedin** [comment]
> Also, wayland support.

Official binary supports wayland, just use `QT_QPA_PLATFORM=wayland`, just like with any other Qt app

**Comment by portaloffreedom** [comment]
@ilya-fedin I'm sorry if I reacted a bit too harsh but I wanted to make it clear what the problem for half of the people was here.

All of this is old stuff back when it was a problem. Now it's not any more and if package maintainers have problems again with something similar they can file a new bug or reopen this one.
Given everything that I've said I think this bug can be safely closed now.

**Comment by ilya-fedin** [comment]
> To sum it up: the official binary can include the Qt libraries, I don't think people have problems with that. But please don't fork Qt for useless features so that it would only make the application 10 times bigger in our computers for the sake of it, while removing compatibility for new stuff e.g. with new wayland.
> 
> Make your life easier, leave Qt development to upstream.

Qt patches never were a barrier for this. Actually, on linux patches are used and using to backport fixes from newer Qt. The barriers were that build system had hardcoded paths to static libraries and Qt private APIs usage by tdesktop. The first is fixed by migrating fro gyp to cmake, the second.. Well, it is not fixed and I have a feel that this is impossible without the full rewrite of tdesktop, like @john-preston said, but it is workarounded by including some parts of Qt in tdesktop from distropatches

**Comment by emanuelserpa** [comment]
> > I think that the issue was about compiling the Telegram's Desktop
> 
> tdesktop imported distropatches in January, it can be compiled with system Qt in most of distros

Yeah, looks like everybody used downstream patches in January still

https://github.com/archlinux/svntogit-community/commit/23eff2b1ef7435441e93120618ca899f0b0e7e61#diff-91e904bd64036c9c236ce84542f97781

Now: https://github.com/archlinux/svntogit-community/blob/packages/telegram-desktop/trunk/PKGBUILD

Can I ask if you guys still use the downstream qt for the official binary?

**Comment by ilya-fedin** [comment]
> archlinux/svntogit-community@23eff2b#diff-91e904bd64036c9c236ce84542f97781

The next commit xd

https://github.com/archlinux/svntogit-community/commit/0917e4882d1155c3413ae6ade16d1c6669527497#diff-91e904bd64036c9c236ce84542f97781

> Can I ask if you guys still use the downstream qt for the official binary?

Probably my previous message answers that

**Comment by emanuelserpa** [comment]
So, the second part (I don't know what it is) is "fixed" by a patch that workaround the "limitations" of the mainline qt, right? 

Then the fix is officially compiling with the patches and moving on.

**Comment by ilya-fedin** [comment]
> So, the second part (I don't know what it is) is "fixed" by a patch that workaround the "limitations" of the mainline qt, right?

By adding the qt_functions.cpp from distropatches that conatins private functions that is not exposed by Qt libs

> Then the fix is officially compiling with the patches and moving on.

What do you mean? :hmm:
